File: src/com/drawmetry/pomodoro/PomodoroTimer.java
package com.drawmetry.pomodoro;

import java.applet.Applet;
import java.applet.AudioClip;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.net.URL;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JSpinner;
import javax.swing.JTextField;
import javax.swing.JToolBar;
import javax.swing.SpinnerModel;
import javax.swing.SpinnerNumberModel;
import javax.swing.SwingUtilities;
import javax.swing.Timer;

public class PomodoroTimer implements Runnable, ActionListener {

	private Timer ticker = new Timer(1000, this);
	private long endTime;
	private long timeRemaining = 0L;
	private JButton startCancelButton;
	private boolean running;
	private SpinnerModel hourModel = new SpinnerNumberModel(0, 0, 12, 1);
	private SpinnerModel minuteModel = new SpinnerNumberModel(25, 0, 59, 1);
	private SpinnerModel secondModel = new SpinnerNumberModel(0, 0, 59, 1);
	private JTextField timeField;

	public static void main(String[] args) {
		SwingUtilities.invokeLater(new PomodoroTimer());
	}

	@Override
	public void run() {
		JFrame frame = new JFrame("Pomodoro");
		frame.setLayout(new BorderLayout());
		JToolBar toolBar = configToolBar();
		frame.add(toolBar, BorderLayout.NORTH);
		configTimeField();
		frame.add(timeField, BorderLayout.CENTER);
		configStartCancelButton();
		frame.add(startCancelButton, BorderLayout.SOUTH);
		frame.pack();
		frame.setResizable(false);
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frame.setVisible(true);

	}

	@Override
	public void actionPerformed(ActionEvent e) {
		timeRemaining = endTime - System.currentTimeMillis();
		if (timeRemaining <= 0L) {
			timeRemaining = 0L;
			startCancelButton.setText("START");
			running = false;
			ticker.stop();
			buzz();
		}
		timeField.setText(formatTime(timeRemaining));
	}

	private JToolBar configToolBar() {
		JToolBar toolBar = new JToolBar();
	
		toolBar.add(new JLabel(" H:"));
		JSpinner hourSpinner = new JSpinner(hourModel);
		toolBar.add(hourSpinner);
	
		toolBar.add(new JLabel(" M:"));
		JSpinner minuteSpinner = new JSpinner(minuteModel);
		toolBar.add(minuteSpinner);
	
		toolBar.add(new JLabel(" S:"));
		JSpinner secondSpinner = new JSpinner(secondModel);
		toolBar.add(secondSpinner);
		return toolBar;
	}

	private void configTimeField() {
		timeField = new JTextField("0:00:00");
		timeField.setBackground(Color.RED);
		timeField.setForeground(Color.WHITE);
		timeField.setFont(new Font("Helvetica", Font.PLAIN, 48));
		timeField.setHorizontalAlignment(JTextField.CENTER);
	}

	private void configStartCancelButton() {
		startCancelButton = new JButton("START");
		startCancelButton.addActionListener(new ActionListener() {
	
			@Override
			public void actionPerformed(ActionEvent e) {
				startCancelButtonClicked();
			}
	
		});
	}

	private void startCancelButtonClicked() {
		if (!running) {
			start();
		} else {
			cancel();
		}
	}

	private void start() {
		long now = System.currentTimeMillis();
		endTime = now + getTime();
		startCancelButton.setText("CANCEL");
		running = true;
		ticker.start();
	}

	private void cancel() {
		ticker.stop();
		timeRemaining = 0L;
		startCancelButton.setText("START");
		running = false;
		timeField.setText(formatTime(timeRemaining));
	}

	private long getTime() {
		long h = ((Integer) hourModel.getValue()).longValue();
		long m = ((Integer) minuteModel.getValue()).longValue();
		long s = ((Integer) secondModel.getValue()).longValue();
		return (((h * 60L) + m) * 60L + s) * 1000L;
	}

	private void buzz() {
		URL url = getClass().getResource("res/shipsbell.wav");
		AudioClip clip = Applet.newAudioClip(url);
		clip.play();
	}

	private String formatTime(long time) {
		long seconds = (time + 500) / 1000; // round to nearest second
		long minutes = seconds / 60;
		long hours = minutes / 60;
		return String.format("%d:%02d:%02d", hours, minutes % 60, seconds % 60);
	}

}

File: src/com/drawmetry/pomodoro/PomodoroTimer.kt
package com.drawmetry.pomodoro

import java.applet.Applet
import java.awt.Color
import java.awt.Font
import java.awt.GridLayout
import java.awt.event.ActionEvent
import javax.swing.*
import java.lang.System.currentTimeMillis as now

class PomodoroTimer : Runnable {

    private var endTime: Long = 0L
    private val running: Boolean
        get() = now() < endTime

    private val timeBar = PomodoroToolBar()

    private val timeField: JTextField by lazy {
        val jTextField = JTextField("0:00:00")
        jTextField.isEditable = false
        jTextField.background = Color.RED
        jTextField.foreground = Color.WHITE
        jTextField.font = Font("Helvetica", Font.PLAIN, 24)
        jTextField.horizontalAlignment = JTextField.CENTER
        jTextField
    }

    private val startCancelButton: JButton by lazy {
        val button = JButton("START")
        button.addActionListener {
            if (!running) {
                start()
            } else {
                cancel()
            }
        }
        button
    }

    private val ticker: Timer = Timer(1000, ::onTick)

    override fun run() {
        val frame = JFrame("Pomodoro")
        frame.layout = GridLayout(1, 3, 0, 0)
        frame.add(timeBar)
        frame.add(timeField)
        frame.add(startCancelButton)
        frame.pack()
        frame.isResizable = false
        frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
        frame.isVisible = true
    }

    private fun start() {
        val time = timeBar.time
        endTime = now() + time
        timeField.text = time.toTimeString()
        startCancelButton.text = "CANCEL"
        ticker.start()
    }

    private fun cancel() {
        endTime = 0L
        timeField.text = 0L.toTimeString()
        startCancelButton.text = "START"
        ticker.stop()
    }

    private fun onTick(@Suppress("UNUSED_PARAMETER") e: ActionEvent) {
        val timeRemaining = endTime - now()
        if (timeRemaining > 0L) {
            timeField.text = timeRemaining.toTimeString()
        } else {
            startCancelButton.text = "START"
            ticker.stop()
            buzz()
            timeField.text = 0L.toTimeString()
        }
    }

    private fun buzz() {
        val url = javaClass.getResource("res/shipsbell.wav")
        val clip = Applet.newAudioClip(url)
        clip.play()
    }

    private fun Long.toTimeString(): String {
        val seconds = (this + 500L) / 1000L // round to nearest second
        val minutes = seconds / 60L
        val hours = minutes / 60L
        return "%d:%02d:%02d".format(hours, minutes % 60L, seconds % 60L)
    }

    // Need this so that this class can be declared as main class in the
    // jar manifest of the generated runnable jar.
    companion object {

        @JvmStatic
        fun main(vararg args: String) {
            SwingUtilities.invokeLater(PomodoroTimer())
        }
    }
}


